paddle/framework/net_design.md

```diff
@@ -0,0 +1,174 @@
+  // `ops_` to run, If no positive indexes are provided, all operators in `ops_`
+  // will run.
+  virtual bool Run(Scope *scope, OpIndex begin = -1,
+                   OpIndex end = -1) const = 0;
+};
+```
+
+All network implementations should build networks from a protobuf message which 
+describes the structure of a real network; `Run` method should be implemented by 
+all implementations to offer a universal method to forward or backward compute a network.
+
+A method of factory pattern can be defined like
+
+```c++
+std::unique<NetworkBase> CreateNet(const NetDef& def) {
+  switch (def.model_type()) {
+    case NN:
+      return new Network(def);
+    case Recursive:
+      return new RecursiveNet(def);
+    case Recurrent:
+      return new RecurrentNet(def);
+  }
+  return nullptr;
+}
+```
+
+Network is designed as the container of operators, to make it more extendable,
+we decoupling it from the related variable resources. 
+
+`Run(Scope* scope)` takes the scope as a argument so that it can run in different scopes.
+
+Finally, `NetworkBase` can be used as followed
+
+```c++
+Scope default_scope;
+auto net = CreateNet(def);
+
+if (net) {
+  net.Run(&default_scope);
+}
+```
+
+
+## A Simple Network Implementation
+
+A very basic implementation is as followed, all it does is simply to run every operators in sequence.
+
+```c++
+class ScratchNet final : public NetworkBase {
+ public:
+  // Create a network describe by `def`.  NetDef is the definition of a network.
+  ScratchNet(const NetDef &def);
+
+  virtual bool InferShape(Scope *scope) override;
+
+  // Run all the operators with the `scope`, if no scope is provided, default
+  // scope will be used instead.
+  virtual bool Run(Scope *scope = nullptr, OpIndex begin,
+                   OpIndex end) const override;
+
+  const std::vector<Operator> &GetOps() const;
+
+  std::vector<Operator> *MutableOps();
+
+ protected:
+  // Create operators accordding to `def`.
+  bool CreateNet(const NetDef &def);
+
+  // Add a operator which is identified as `type` and has attributes described
+  // in `attrs`, the `inputs` are the keys of readonly input variables,
+  // `outputs` are keys of mutable output variables. An `OpIndex` will be
+  // returned which indicates the offset of the new operator in `ops_`.
+  OpIndex AddOp(const std::string &type, const std::vector<string> &inputs,
+                const std::vector<string> &outputs,
+                const OprAttr &attrs = OprAttr());
+
+ private:
+  // the operators owned by `Network`.
+  std::vector<Operator> ops_;
+};
+```
+
+`ScratchNet` will create operators so that a private member `ops_` is defined,
+the operators are created by `CreateNet`, and each operator is created by `AddOp`.
+
+
+## Usage
+`ScratchNet` can be used to define and run a network as followed
+
+```c++
+// create an empty scope located on CPU device.
+Scope scope(CPUPlace());
+
+// create and init variables described in `net_desc`.
+scope.CreateVariables(net_desc);
```

Comment:
`Variable` information is not stored in `NetDef`, and in `OpDef`, there will only be input/output names for `Variable`. 

I think `CreateVariables` is a high-level API, and should not be added to `scope`. Maybe there should be a high-level concept called `NetworkBuilder` or `Model`. A `NetworkBuilder` takes a `NetworkBase` and a `Scope` as its inputs. In `NetworkBuilder`, user can add layers, add backward ops, and get all parameters.

```cpp
NetworkBase network;
Scope scope;
Variable* image = scope.CreateVariable("Image");
Variable* label = scope.CreateVariable("Label");
NetworkBuilder builder(&network, &scope);

Variable* fc_out = builder.FCLayer(input=image, size=100, activation="Sigmoid");
Variable* prediction = builder.FCLayer(input=fc_out, size=10, activation="Sigmoid");
Variable* loss = builder.CrossEntropy(input=prediction, label=label);
Variable* avg_loss = builder.Mean(loss);

auto allParams = builder.Parameters();
builder.BackwardFrom(avg_loss)
builder.AddOptimization(1e-4, "adam");

// train one mini-batch
network.run(&scope);
```